### PR TITLE
bpo-34002: Minor efficiency and clarity improvements in email package.

### PR DESCRIPTION
**Pull request by selik**
Creating a copy of a list, then repeatedly popping the first element could be much slower than necessary if the list is long. Further explanations in the commit messages

Automerge-Triggered-By: @maxking

### CODE REVIEW

File: Lib/email/headerregistry.py
Comment (serhiy-storchaka) [inline Lib/email/headerregistry.py]: This can be written as
```python
if disp is not None and not parser.SPECIALS.isdisjoint(disp):
```

Comment (serhiy-storchaka) [reply Lib/email/headerregistry.py]: There is also a similar code at line 90.

File: Lib/email/utils.py
Comment (serhiy-storchaka) [inline Lib/email/utils.py:262]: What if `params[0]` is not a 2-tuple?

**Review by serhiy-storchaka — COMMENTED** [review]
Please open an issue on the tracker for discussing this PR.

File: Lib/email/utils.py
Comment (selik) [reply Lib/email/utils.py:262]: That's possible for the first params element, though the others are implicitly checked by the unpacking for-loop.

Looking at the usage of `decode_params`, it's only use is in `email.message._get_params_preserve` which in turn has 3 call sites. All 3 calls unpack the elements as `(k, v)` pairs. The bad data would cause an immediate error, given how the method is used. The error would be the same as today, though the traceback would not be as deep.

If that's not satisfactory, I'll revert to:

    key, value = params[0]
    new_params = [(key, value)]

File: Lib/email/headerregistry.py
Comment (selik) [reply Lib/email/headerregistry.py]: Updated the code to use ``set.intersection``, which I think reads better than "not isdisjoint". I spotted a 3rd similar comparison and changed that one as well.

I noted in the commit message that this avoids an unnecessary set construction.

**Comment by selik** [comment]
I created an issue in the tracker and changed the name of this pull-request.

https://bugs.python.org/issue34002

File: Lib/email/headerregistry.py
Comment (rhettinger) [inline Lib/email/headerregistry.py]: Consider using `set.isdisjoint()`.

File: Lib/email/headerregistry.py
Comment (serhiy-storchaka) [reply Lib/email/headerregistry.py]: `set.intersection()` creates a new set, while `set.isdisjoint()` returns a boolean and stops fast if a common item is found.

File: Lib/email/headerregistry.py
Comment (serhiy-storchaka) [inline Lib/email/headerregistry.py]: You could use `lp` instead of `self.username`.

File: Lib/email/headerregistry.py
Comment (selik) [reply Lib/email/headerregistry.py]: @serhiy-storchaka also suggested "not isdisjoint". Is that because it explicitly returns a `bool`?

File: Lib/email/utils.py
Comment (serhiy-storchaka) [inline Lib/email/utils.py:267]: `params[1:]` still creates a copy of the list.

Comment (selik) [reply Lib/email/utils.py:267]: True. The real benefit is avoiding repeated `.pop(0)`.

File: Lib/email/headerregistry.py
Comment (selik) [reply Lib/email/headerregistry.py]: GitHub hid @serhiy-storchaka's explanation after I pushed a change. Makes sense to stop early when finding a match via disjoint. I'll make the change ...

File: Lib/email/headerregistry.py
Comment (selik) [reply Lib/email/headerregistry.py]: Fixed.

File: Lib/email/utils.py
Comment (serhiy-storchaka) [reply Lib/email/utils.py:267]: But the code can be made more optimal if avoid a copy.

Comment (terryjreedy) [reply Lib/email/utils.py:267]: It looks to me that the final value of 'encoded' only depends on the final value of name.  If so, the loop is equivalent to
```
if params:
    encoded = params[-1][0].endswith('*')
```
Otherwise, one can skip-iterate without copying with
```
it = iter(params)
try:
    next(it)
except StopIteration:
    pass
for name, value in it:
   ...
```

**Comment by maxking** [comment]
This PR seems to be a good and looks like has reviews from several core devs.
Since there isn't any change of behaviour or bugfix, I don't know if there is a need for NEWS file here. It also looks like @selik has made all the requested changes.

Can @serhiy-storchaka, @terryjreedy @warsaw @bitdancer merge this with label to skip news entry? Unless we do do need a NEWS entry, in which case it would be good to know that.

**Comment by terryjreedy** [comment]
Though sometimes disagreeing on meaning of 'trivial', we generally want an issue and news entry for non-trivial changes, and this is non-trivial.  Someone should add a short news entry.  It will inform anyone looking for email changes that might have caused an obscure bug that we made a change that we *think* should have no visible effect other than performance.

**Comment by maxking** [comment]
Thanks @terryjreedy, that totally makes sense.

@selik can you please add a new entry for this PR?

**Comment by maxking** [comment]
I am okay with the changes, like I mentioned before, but like @terryjreedy mentioned, it would be good to have a NEWS entry. 

I'd push a NEWS entry myself if I had access to push to his PR (which can be done using the flag "Allow changes from maintainers flag", I think.

**Comment by selik** [comment]
I believe the box is checked to allow edits from maintainers

**Comment by terryjreedy** [comment]
@maxking I am able to edit the files, so you should be able to create a branch, add a blurb, and push.

**Review by maxking — APPROVED** [review]
LGTM

**Comment by maxking** [comment]
Thanks @selik for this :)

File: Lib/email/utils.py
Comment (selik) [reply Lib/email/utils.py:267]: Each iteration of the loop uses the `encoded` value. The `iter` and `next`ing would be a better technique.

```
it = iter(params)
new_params = [next(it)]
for name, value in it:
    ...
```

**Pull request merged**
